pandas/core/index.py

```diff
@@ -391,6 +395,11 @@ def _shallow_copy(self, values=None, infer=False, **kwargs):
 
         if infer:
             attributes['copy'] = False
+            if self._infer_as_myclass:
```

Comment:
Ah I see. One idea is splitting the method. Based on the current impl:

- Use ``_simple_new`` when inference is not needed.
- Use ``_shallow_copy`` when inference is needed.

I've once tried it, but it isn't work by the simple change because of some impl differences, for example ``MultiIndex`` doesn't have ``_simple_new``.

Let me try it in a separate PR, or change the milestone to 0.18.0.